websocket/polygon.go

```diff
@@ -0,0 +1,228 @@
+	rQueue chan []byte
+	wQueue chan []byte
+
+	log Logger
+}
+
+// todo: might want to separate Connect logic out from New
+func New(config Config) (*Client, error) {
+	if config.APIKey == "" {
+		return nil, errors.New("API key is required")
+	}
+
+	url := fmt.Sprintf("wss://%v.polygon.io/%v", string(config.Feed), string(config.Market))
+
+	if config.Log == nil {
+		config.Log = zap.NewNop().Sugar()
+	}
+
+	// todo: is this default dialer sufficient? might want to let user pass in a context so they can cancel the dial
+	conn, res, err := websocket.DefaultDialer.Dial(url, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to dial server: %w", err)
+	} else if res.StatusCode != 101 {
+		return nil, errors.New("server failed to switch protocols")
+	}
+
+	conn.SetReadLimit(maxMessageSize)
+	if err := conn.SetReadDeadline(time.Now().Add(pongWait)); err != nil {
+		return nil, fmt.Errorf("failed to set read deadline: %w", err)
+	}
+	conn.SetPongHandler(func(string) error {
+		return conn.SetReadDeadline(time.Now().Add(pongWait))
+	})
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	c := &Client{
+		apiKey: config.APIKey,
+		ctx:    ctx,
+		cancel: cancel,
+		conn:   conn,
+		rQueue: make(chan []byte, 10000),
+		wQueue: make(chan []byte, 100),
+		log:    config.Log,
+	}
+
+	go c.read()
+	go c.write()
+	go c.process()
+
+	return c, nil
+}
+
+// todo: Subscribe, Unsubscribe, etc
+
+func (c *Client) Close() error {
+	c.cancel()
+	err := c.conn.WriteControl(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""), time.Now().Add(writeWait))
+	if err != nil {
+		c.log.Errorf("failed to gracefully close: %v", err)
+		return err
+	}
+	c.log.Infof("connection closed successfully")
+	return nil
+}
+
+func (c *Client) read() {
+	defer func() {
+		c.log.Debugf("closing read thread")
+		c.conn.Close() // todo: should this force close?
+	}()
+
+	for {
+		select {
+		case <-c.ctx.Done():
+			return
+		default:
+			_, msg, err := c.conn.ReadMessage()
+			if err != nil {
+				if websocket.IsCloseError(err, websocket.CloseNormalClosure) {
+					break
+				} else if websocket.IsUnexpectedCloseError(err, websocket.CloseNormalClosure) {
+					c.log.Errorf("connection closed unexpectedly: %v", err)
+					break
+				}
+				c.log.Errorf("failed to read message: %v", err)
+				break
+			}
+			c.rQueue <- msg
+		}
+	}
+}
+
+func (c *Client) write() {
+	ticker := time.NewTicker(pingPeriod)
+	defer func() {
+		c.log.Debugf("closing write thread")
+		ticker.Stop()
+		c.conn.Close() // todo: should this force close?
+	}()
+
+	for {
+		select {
+		case <-c.ctx.Done():
+			return
+		case <-ticker.C:
+			if err := c.conn.SetWriteDeadline(time.Now().Add(writeWait)); err != nil {
+				c.log.Errorf("failed to set write deadline: %v", err)
+				return // todo: should this return?
+			}
+			if err := c.conn.WriteMessage(websocket.PingMessage, nil); err != nil {
+				c.log.Errorf("failed to send ping message: %v", err)
+				return
+			}
```

Comment:
I think you can use `conn.WriteControl` with the `PingMessage` type here to simplify this into one function call